mongocrypt/src/ctx.rs

```diff
@@ -0,0 +1,686 @@
+        }
+        Some(KmsCtx {
+            inner,
+            _marker: PhantomData,
+        })
+    }
+}
+
+impl<'ctx> Drop for KmsScope<'ctx> {
+    fn drop(&mut self) {
+        unsafe {
+            // If this errors, it will show up in the next call to `ctx.status()` (or any other ctx call).
+            sys::mongocrypt_ctx_kms_done(*self.ctx.inner.borrow());
+        }
+    }
+}
+
+/// Manages a single KMS HTTP request/response.
+pub struct KmsCtx<'scope> {
+    inner: *mut sys::mongocrypt_kms_ctx_t,
+    _marker: PhantomData<&'scope mut ()>,
+}
+
+unsafe impl<'scope> Send for KmsCtx<'scope> {}
+unsafe impl<'scope> Sync for KmsCtx<'scope> {}
+
+impl<'scope> HasStatus for KmsCtx<'scope> {
+    unsafe fn native_status(&self, status: *mut sys::mongocrypt_status_t) {
+        sys::mongocrypt_kms_ctx_status(self.inner, status);
+    }
+}
+
+impl<'scope> KmsCtx<'scope> {
+    /// Get the HTTP request message for a KMS handle.
+    pub fn message(&self) -> Result<&'scope [u8]> {
+        // Safety: the message referenced has a lifetime that's valid until kms_done is called,
+        // which can't happen without ending 'scope.
+        unsafe {
+            let bin = Binary::new();
+            if !sys::mongocrypt_kms_ctx_message(self.inner, *bin.native()) {
+                return Err(self.status().as_error());
+            }
+            bin.bytes()
+        }
+    }
+
+    /// Get the hostname from which to connect over TLS.
+    ///
+    /// The endpoint consists of a hostname and port separated by a colon.
+    /// E.g. "example.com:123". A port is always present.
+    pub fn endpoint(&self) -> Result<&'scope str> {
+        let mut ptr: *const ::std::os::raw::c_char = ptr::null();
+        unsafe {
+            if !sys::mongocrypt_kms_ctx_endpoint(
+                self.inner,
+                &mut ptr as *mut *const ::std::os::raw::c_char,
+            ) {
+                return Err(self.status().as_error());
+            }
+            Ok(CStr::from_ptr(ptr).to_str()?)
+        }
+    }
+
+    /// Indicates how many bytes to feed into `feed`.
+    pub fn bytes_needed(&self) -> u32 {
+        unsafe { sys::mongocrypt_kms_ctx_bytes_needed(self.inner) }
+    }
+
+    /// Feed bytes from the HTTP response.
+    ///
+    /// Feeding more bytes than what has been returned in `bytes_needed` is an error.
+    pub fn feed(&mut self, bytes: &[u8]) -> Result<()> {
+        let bin = BinaryRef::new(bytes);
+        unsafe {
+            if !sys::mongocrypt_kms_ctx_feed(self.inner, *bin.native()) {
+                return Err(self.status().as_error());
+            }
+        }
+        Ok(())
+    }
+
+    /// Get the KMS provider identifier associated with this KMS request.
+    ///
+    /// This is used to conditionally configure TLS connections based on the KMS
+    /// request. It is useful for KMIP, which authenticates with a client
+    /// certificate.
+    ///
+    /// Returns one of the static strings: "aws", "azure", "gcp", or "kmip".
```

Comment:
Yup, done.  This also made me rethink how I was handling unexpected values for the other enums, so now all of them have a catchall rather than immediately erroring; this'll be better for forward compatibility.

Reply:
Just ran across [this](https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#why-are-extraoptions-and-kmsproviders-maps) documentation, which says that we don't want AWS as part of the public types, so I've reverted this back to a string.  I kept the fallbacks for other enums.

Reply:
Hmm, we already have AWS show up in our [auth mechanisms](https://docs.rs/mongodb/latest/mongodb/options/enum.AuthMechanism.html#variant.MongoDbAws), so this guidance seems to be a bit inconsistent. I wonder if we could ask for clarification on this? It's possible it means that we don't want to be modelling AWS APIs via our public API (i.e. have rigid structs for possibly changing AWS provider info) rather than any sort of general brand issue.

Reply:
Turns out that documentation was stale, so this is back to an enum.